crates/static-analysis-kernel/src/analysis/ddsa_lib/runtime.rs

```diff
@@ -387,6 +433,101 @@ for (const queryMatch of globalThis.__RUST_BRIDGE__query_match) {{
+        loop {
+            let mut state = cvar
+                .wait_while(lock.lock().unwrap(), |state| {
+                    state.execution_complete && !state.should_shut_down
+                })
+                .expect("mutex should not be poisoned");
+
+            if state.should_shut_down {
+                break;
+            }
+
+            // Any instant after `timeout_threshold` will trigger the timeout
+            let timeout_threshold = state.start_time + state.timeout_duration;
+            let now = Instant::now();
+
+            if now >= timeout_threshold {
+                // This branch represents an edge case where the OS couldn't wake this thread up
+                // until after the watchdog should've already triggered a timeout.
+
+                // Trust that v8 will halt execution and eagerly set the state to complete so the
+                // main thread doesn't need to acquire the lock to do it.
+                state.execution_complete = true;
+                drop(state);
+                v8_handle.terminate_execution();
+            } else {
+                // This is guaranteed not to underflow
+                let additional_wait = timeout_threshold - now;
+                let result = cvar
+                    .wait_timeout_while(state, additional_wait, |state| !state.execution_complete)
+                    .expect("mutex should not be poisoned");
+                state = result.0;
+
+                // If the condvar triggered a timeout, `execution_complete` _must_ be false, because of
+                // our use of `wait_timeout_while`. Thus, it's always appropriate to terminate execution.
+                if result.1.timed_out() {
+                    // Trust that v8 will halt execution and eagerly set the state to complete so the
+                    // main thread doesn't need to acquire the lock to do it.
+                    state.execution_complete = true;
+                    drop(state);
+                    v8_handle.terminate_execution();
+                }
+            }
+        }
+    });
+    state_pair
+}
+
+/// A struct used to communicate state from a `JsRuntime` to a watchdog thread that calls
+/// [`v8::Isolate::terminate_execution`]. To request the watchdog to enforce a timeout,
+/// this struct should be populated with `timeout_duration`, `start_time`, and `execution_complete`
+/// should be set to `false`. After that, the watchdog thread should be notified via a condvar.
+///
+/// `should_shut_down` should only be set to `true` when the `JsRuntime` is being dropped.
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+struct JsExecutionState {
+    timeout_duration: Duration,
+    start_time: Instant,
+    execution_complete: bool,
+    should_shut_down: bool,
+}
+
+impl JsExecutionState {
+    pub fn new() -> Self {
+        Self {
+            timeout_duration: Duration::default(),
+            start_time: Instant::now(),
+            execution_complete: true,
```

Comment:
When we create the state, the execution is not complete. Why setting it to `true` then?